refactor: make vault poolManager agnostic

chefburger opened the pull request:
According our original design, our vault contract should be general enough to support not only AMM pools but also any other types of applications on the top. This PR aims to make the vault poolManager agnostic, so that we are not limited to AMM poolManagers.

**1. Main change: removed `PoolKey` & `BalanceDelta` from `Vault.sol` which is AMM specific**
**2. Drawback: gas cost went up due to the extra call to `accountPoolBalanceDelta`**

Tried this but gas cost went up even more 
```
    function accountPoolBalanceDelta(Currency[] calldata currencys, int128[] calldata deltas, address settler)
        external
        override
        isLocked
        onlyPoolManager
    {
        // TODO: // require(currencys.length == deltas.length, "Vault: invalid input length");

        for (uint256 i = 0; i < currencys.length; i++) {
            // keep track on each pool manager
            _accountDeltaOfPoolManager(msg.sender, currencys[i], deltas[i]);

            // keep track of the balance for the whole vault
            SettlementGuard.accountDelta(settler, currencys[i], deltas[i]);
        }
    }
```


Thinking not merge until we find better way ?

ChefSnoopy commented:
https://github.com/pancakeswap/pancake-v4-core/commit/d64c44c7f21b7c7864d6ab0e2e86539576076e2a

I updated a new version as below, lower gas than this version, but still increases too much gas cost than the original.
...................................................................................................................
function accountPoolBalanceDelta(Currency currency0, Currency currency1, BalanceDelta delta, address settler)

chefburger commented:
> version two commit: [d64c44c](https://github.com/pancakeswap/pancake-v4-core/commit/d64c44c7f21b7c7864d6ab0e2e86539576076e2a)
> 
> I updated a new version as below, lower gas than this version, but still increases too much gas cost than the original. ................................................................................................................... function accountPoolBalanceDelta(Currency currency0, Currency currency1, BalanceDelta delta, address settler)



> version two commit: [d64c44c](https://github.com/pancakeswap/pancake-v4-core/commit/d64c44c7f21b7c7864d6ab0e2e86539576076e2a)
> 
> I updated a new version as below, lower gas than this version, but still increases too much gas cost than the original. ................................................................................................................... function accountPoolBalanceDelta(Currency currency0, Currency currency1, BalanceDelta delta, address settler)

Thinking we should not keep `BalanceDelta` as well because the data type has an assumption that there are 2 currencies involved but that might no be the case

chefburger commented:
> > Thinking we should not keep `BalanceDelta` as well because the data type has an assumption that there are 2 currencies involved but that might no be the case
> 
> Would this be something we can consider? given that gas usage is highly critical for our dex use cases
> 
> 1/ keeping our current `accountPoolBalanceDelta(PoolKey, BalanceDelta, settler)` 2/ include this generic balance as another method `accountCurrencyBalanceDelta(Currency, int128 delta, settler)`

Cool, that make sense to me..  wdyt @ChefSnoopy

chefburger commented:
Thx for advices from Mist, new update involved:

1. removed `IPoolManager` from `Vault.sol`, because it's AMM specific
2. aded a general version `accountPoolBalanceDelta` for future non-AMM poolManager usage
3. simplified `onlyPoolManager` check logic

chef-omelette commented on src/Vault.sol:
```diff
@@ -24,16 +23,11 @@ contract Vault is IVault, VaultToken, Ownable {
-        /// @dev Make sure:
-        /// 1. the pool manager specified in PoolKey is the caller
-        /// 2. the pool manager has been registered
-        if (poolManager != msg.sender) revert NotFromPoolManager();
```
to be more precise in semantic, I think we can rename all `poolManager` to `registeredContract`. what do you think?

chefburger replied:
how 
1. `registerPoolManager` => `registerApp`
2. `onlyPoolManager` => `onlyRegisteredApp`
3. `reservesOfPoolManager` => `reservesOfApp`

chef-omelette replied:
lgtm!

ChefMist replied:
agree as well.. will review once that change is applied

chef-omelette reviewed (dismissed):
LGTM

chef-omelette commented:
> > Thinking we should not keep `BalanceDelta` as well because the data type has an assumption that there are 2 currencies involved but that might no be the case
> 
> Would this be something we can consider? given that gas usage is highly critical for our dex use cases
> 
> 1/ keeping our current `accountPoolBalanceDelta(PoolKey, BalanceDelta, settler)` 2/ include this generic balance as another method `accountCurrencyBalanceDelta(Currency, int128 delta, settler)`

upvote for this

ChefMist commented on src/interfaces/IVault.sol:
```diff
@@ -30,16 +26,16 @@ interface IVault is IVaultToken {
 
-    /// @notice enable or disable specific pool manager
-    function registerPoolManager(address poolManager) external;
+    /// @notice register a app so that it can perform accounting base on vault
```
```suggestion
    /// @notice register an app so that it can perform accounting base on vault
```
